x/aswap/handler.go

```diff
@@ -0,0 +1,307 @@
+package aswap
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/migration"
+	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/x/cash"
+	"github.com/tendermint/tendermint/libs/common"
+)
+
+const (
+	// pay swap cost up-front
+	createSwapCost  int64 = 300
+	returnSwapCost  int64 = 0
+	releaseSwapCost int64 = 0
+
+	// currently set to two days
+	day = time.Hour * 24
+	// amount of days for minTimeout
+	timeoutDays = 2
+	minTimeout  = timeoutDays * day
+
+	tagSwapId string = "swap-id"
+	tagAction string = "action"
+)
+
+// RegisterRoutes will instantiate and register
+// all handlers in this package
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, cashctrl cash.Controller) {
+	bucket := NewBucket()
+
+	r.Handle(pathCreateSwap, migration.SchemaMigratingHandler("aswap", CreateSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReleaseSwap, migration.SchemaMigratingHandler("aswap", ReleaseSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReturnReturn, migration.SchemaMigratingHandler("aswap", ReturnSwapHandler{auth, bucket, cashctrl}))
+}
+
+// RegisterQuery will register this bucket as "/aswaps"
+func RegisterQuery(qr weave.QueryRouter) {
+	NewBucket().Register("aswaps", qr)
+}
+
+//---- create
+
+// CreateSwapHandler creates a swap
+type CreateSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.CoinMover
+}
+
+var _ weave.Handler = CreateSwapHandler{}
+
+// Check does the validation and sets the cost of the transaction
+func (h CreateSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &weave.CheckResult{
+		GasAllocated: createSwapCost,
+	}
+	return res, nil
+}
+
+// Deliver moves the tokens from sender to the swap account if all conditions are met.
+func (h CreateSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a swap object
+	swap := &Swap{
+		Metadata:     &weave.Metadata{},
```

Comment:
Shouldn't we set the schema here? Based on the schema of the msg?

Reply:
In fact this code should also fail if `msg.Metadata.Schema != 1`, so it rejects "future schemas"

Reply:
@husio please correct me if i am wrong here

Reply:
yeah, that will complain during the validation in tests, will be fixed after I've tested everything